### PR TITLE
feature/packedbubble-verlet

### PR DESCRIPTION
**Pull request by Izothep**
Extension of packed bubble series based on Verlet integration.

Old version (now called spiral-packing):
https://highcharts.github.io/highcharts-utils/samples/#gh/28fb76fdf2/sample/highcharts/demo/packed-bubble-spiral
Image:
Screenshot 2019-03-12 at 17 56 42

Was not recalculating the bubble positions after initial creation. 

With new algorithm **forces approach** was implemented, so all the bubbles may interfere with each other, resulting in more complicated layouts. 
https://highcharts.github.io/highcharts-utils/samples/#gh/28fb76fdf2/sample/highcharts/demo/packed-bubble

Image:
Screenshot 2019-03-12 at 17 56 18

On top of new algorithm, the next extension was developed, allowing series to split into independent groups. These is giving a chance for more aggregated graphs creation, where the comparison between series is simpler.
https://highcharts.github.io/highcharts-utils/samples/#gh/28fb76fdf2/sample/highcharts/demo/packed-bubble-split

Image: 
Screenshot 2019-03-12 at 17 59 43

Done:

- [x] Initial refactoring
- [x] Initial Bugs fixing
- [x] Demo creation
- [x] API Docs

To do: 
- [x] Documentation article
- [x] Gathering feedback and bugs fixing

### CODE REVIEW

**Comment by Izothep** [comment]
https://docs.google.com/document/d/1GkwDNxTVZ8eMfVp4dhVaEgW9dCi1RFlhQeqXCc2Vli0/edit# docs created

File: js/parts-more/PackedBubbleSeries.js [hunk 1]
```diff
@@ -12,82 +12,455 @@ import '../parts/Axis.js';
...
+             * without velocity:
+             * `newPosition += previousPosition - newPosition`.
+             *
+             *
+             *
+             * @since       7.1.0
+             * @sample      highcharts/series-networkgraph/forces/
+             * @validvalue  ["packedbubble"]
+             */
+            integration: 'packedbubble',
```
Comment (pawelfus) [inline js/parts-more/PackedBubbleSeries.js]: Do we want to give users an opportunity to change that? Maybe this option should be `@ ignore`'d?

File: js/parts-more/PackedBubbleSeries.js [hunk 2]
```diff
@@ -12,82 +12,455 @@ import '../parts/Axis.js';
...
+                initialPositionRadius: 100,
+                seriesInteraction: true
+            },
+            enableSimulation: true,
+            /**
+             * Type of the algorithm used when positioning bubbles.
+             *
+             * @validvalue  ["packedbubble"]
+             */
+            type: 'packedbubble',
```
Comment (pawelfus) [inline js/parts-more/PackedBubbleSeries.js]: Do we want to give users an opportunity to change that? Maybe this option should be `@ ignore`'d?

File: js/parts-more/PackedBubbleSeries.js [hunk 3]
```diff
@@ -422,87 +1086,243 @@ seriesType(
...
-
             this.radii = radii;
         },
+        // Draggable mode:
+        /**
+         * Redraw halo on mousemove during the drag&drop action.
+         *
+         * @param {Highcharts.Point} point The point that should show halo.
+         */
+        redrawHalo: function (point) {
```
Comment (pawelfus) [inline js/parts-more/PackedBubbleSeries.js]: Looks similar to `networkgraph.redrawHalo` - synergy here? :)

File: js/parts-more/PackedBubbleSeries.js [hunk 4]
```diff
@@ -422,87 +1086,243 @@ seriesType(
...
+                    newPlotY = point.fixedPosition.plotY - diffY;
+
+                    if (chart.isInsidePlot(newPlotX, newPlotY)) {
+                        point.plotX = newPlotX;
+                        point.plotY = newPlotY;
 
-// When one series is modified, the others need to be recomputed
-H.addEvent(H.seriesTypes.packedbubble, 'updatedData', function () {
-    var self = this;
+                        series.redrawHalo();
```
Comment (pawelfus) [inline js/parts-more/PackedBubbleSeries.js]: This line does nothing - should be `series.redrawHalo(point)`

Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Good catch! It looks that the whole method is not necessary in that case. I have removed it already without any changes in code.

File: js/parts-more/PackedBubbleSeries.js (hunk 2 above)
Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Thanks, changed.

File: js/parts-more/PackedBubbleSeries.js (hunk 1 above)
Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Thanks, changed.

File: js/parts-more/PackedBubbleSeries.js [hunk 5]
```diff
@@ -12,82 +12,455 @@ import '../parts/Axis.js';
...
+        sizeBy: 'area',
         zoneAxis: 'y',
         tooltip: {
             pointFormat: 'Value: {point.value}'
+        },
+        /**
+        * Flag to determine if nodes are draggable or not.
+        * @since 7.1.0
+        */
+        draggable: true,
```
[the remark this answers is not in the document]
Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Comment added.

File: js/parts-more/PackedBubbleSeries.js [hunk 6]
```diff
@@ -12,82 +12,455 @@ import '../parts/Axis.js';
...
+            fillColor: null,
+            lineWidth: 1,
+            lineColor: null,
+            symbol: 'circle'
+        },
+        dataLabels: {
+            formatter: function () {
+                return this.point.value;
+            },
+            allowOverlap: true
```
[the remark this answers is not in the document]
Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Another idea is to use `hideOverlappingLabels` in `series.render` method. In below chart the problem is related to wrongly calculated dataLabels visibility at the beginning of graph, that are not correctly updated. Adding this method seems to do the trick.

File: js/parts-more/PackedBubbleSeries.js [hunk 7]
```diff
@@ -422,87 +1086,243 @@ seriesType(
...
+                    }
+                }
+            }
+        },
+        /**
+         * Mouse up action, finalizing drag&drop.
+         *
+         * @param {Highcharts.Point} point The point that event occured.
+         */
+        onMouseUp: function (point) {
```
[the remark this answers is not in the document]
Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Agree, we should add `redrawHalo` to Mixin as well.

File: js/parts-more/PackedBubbleSeries.js (hunk 3 above)
Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Added as Mixin :)

File: js/parts-more/PackedBubbleSeries.js [hunk 8]
```diff
@@ -12,82 +12,455 @@ import '../parts/Axis.js';
...
+            /**
+             * When initialPositions are set to 'circle', initialPositionRadius
+             * is a distance from the center of circle, in which bubbles are
+             * created.
+             *
+             * @since       7.1.0
+             * @type {Number}
+             * @default 20
+             */
+            initialPositionRadius: 20,
```
[the remark this answers is not in the document]
Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Done. It is influencing the whole graph generation in very interesting way. Definitely worth adding.

File: js/parts-more/PackedBubbleSeries.js [hunk 9]
```diff
@@ -12,82 +12,455 @@ import '../parts/Axis.js';
...
+         * @since   7.1.0
+         * @product highcharts highstock
+         * @default false
+         */
+        useSimulation: false,
+        /**
+         * If parentNode is created for splitted series, parentOption is used
+         * for styling the parent nodes.
+         *
+         * @type    {Object}
```
[the remark this answers is not in the document]
Comment (Izothep) [reply js/parts-more/PackedBubbleSeries.js]: Good idea, done.

**Comment by Izothep** [comment]
Thanks to both of you for comments @pawelfus, @TorsteinHonsi! Commit with all the changes is already pushed.

**Comment by TorsteinHonsi** [comment]
### To do
 - [ ] The `initial-radius` demo is not refered to from a `@sample` tag, meaning it will be inaccessible for the public.
 - [ ] The same for `packed-dashboard`.

**Comment by Izothep** [comment]
Thanks! I have added both demos with series `null` values extremes issue fix.

**Pull request merged**